### DISCUSSION TITLE
help with noob code?

### DISCUSSION THREAD

**Discussion by Web3dGuy** [highfestiva/finplot#132] [Q&A] [answered]
So I've hacked together something that I thought should work, I could use some help finding where I went wrong.
My goal:
download historic price 
download current price
keep historic price and append new prices without redownloading historic
have current live updating candle

currently the solution is to write a temporary csv file with history, then add new lines, read csv into a dataframe, then plot all.
however I feel like I'm feeding data in the wrong way .  one problem is, moving averages and boll band should take the close time and close price for input but when I feed anything other than open time into them I get errors. the other thought I had is that I'm handling the indexing the wrong way, but when I try to index using the time key, more errors.

The program runs , drops a csv,  the candles update , but are messed up and the opens and closes don't line up. 
Then I notice every time the update happens the plot likes to zoom in ? probably due to input problems listed above.

play with my code, tell me all the things I did wrong.

Thanks :)

```python


import pandas as pd
import finplot as fplt
import requests
import json


def get_bars(symbol, interval, limit):
    url = 'https://api.binance.com/api/v1/klines?symbol=' + symbol + '&interval=' + interval + '&limit=' + limit
    data = json.loads(requests.get(url).text)
    cur_data = pd.DataFrame(data)
    cur_data.columns = ['time', 'Open', 'High', 'Low', 'Close', 'Volume', 'Close_time',
                        'qav', 'num_trades', 'taker_base_vol', 'taker_quote_vol', 'ignore']
    cur_data['time'] = pd.to_datetime(cur_data['time'], unit='ms', origin='unix')
    cur_data['Open'] = pd.to_numeric(cur_data["Open"])
    cur_data["Close"] = pd.to_numeric(cur_data["Close"])
    cur_data['High'] = pd.to_numeric(cur_data["High"])
    cur_data['Low'] = pd.to_numeric(cur_data["Low"])
    cur_data['Volume'] = pd.to_numeric(cur_data["Volume"])
    cur_data['Close_time'] = pd.to_datetime(cur_data['Close_time'], unit='ms', origin='unix')
    cur_data['qav'] = pd.to_numeric(cur_data["qav"])
    cur_data['num_trades'] = pd.to_numeric(cur_data["num_trades"])
    cur_data['taker_base_vol'] = pd.to_numeric(cur_data["taker_base_vol"])
    cur_data['taker_quote_vol'] = pd.to_numeric(cur_data["taker_quote_vol"])
    cur_data['ignore'] = pd.to_numeric(cur_data["ignore"])
    cur_data.to_csv('bars.csv', mode='a', header=False, index=False)
    return cur_data


def get_bars_historic(symbol, interval, limit):
    url = 'https://api.binance.com/api/v1/klines?symbol=' + symbol + '&interval=' + interval + '&limit=' + limit
    data = json.loads(requests.get(url).text)
    histdata = pd.DataFrame(data)
    histdata.columns = ['time', 'Open', 'High', 'Low', 'Close', 'Volume', 'Close_time', 'qav', 'num_trades',
                        'taker_base_vol', 'taker_quote_vol', 'ignore']
    histdata['time'] = pd.to_datetime(histdata['time'], unit='ms', origin='unix')
    histdata['Open'] = pd.to_numeric(histdata["Open"])
    histdata["Close"] = pd.to_numeric(histdata["Close"])
    histdata['High'] = pd.to_numeric(histdata["High"])
    histdata['Low'] = pd.to_numeric(histdata["Low"])
    histdata['Volume'] = pd.to_numeric(histdata["Volume"])
    histdata['Close_time'] = pd.to_datetime(histdata['Close_time'], unit='ms', origin='unix')
    histdata['qav'] = pd.to_numeric(histdata["qav"])
    histdata['num_trades'] = pd.to_numeric(histdata["num_trades"])
    histdata['taker_base_vol'] = pd.to_numeric(histdata["taker_base_vol"])
    histdata['taker_quote_vol'] = pd.to_numeric(histdata["taker_quote_vol"])
    histdata['ignore'] = pd.to_numeric(histdata["ignore"])
    histdata.to_csv('bars.csv', header=True, index=False)
    return histdata


def calc_bollinger_bands(df):
    r = df['Close'].rolling(20)
    df['bbh'] = r.mean() + 2*r.std()
    df['bbl'] = r.mean() - 2*r.std()


def update():
    get_cur_data = get_bars('ETHUSDT', '1m', '1')
    current_data = get_cur_data[['time', 'Open', 'Close', 'High', 'Low', 'Volume', 'Close_time']]
    df_total = pd.read_csv('bars.csv').drop_duplicates('time').reset_index(drop=True)
    df_total['time'] = pd.to_datetime(df_total['time'])
    df_total['Close_time'] = pd.to_datetime(df_total['Close_time'])
    print('current data #########################################')
    print(current_data.to_string())
    print('######################################################')
    print('Historic data ########################################')
    print(historic_data.tail().to_string())
    print('######################################################')
    print('Total data ########################################')
    print(df_total.tail().to_string())
    print('######################################################')
    candlesticks = df_total['time Open Close High Low'.split()]
    calc_bollinger_bands(df_total)
    #volumes = DataStruct['time Open Close Volume'.split()
    bollband_hi = df_total['time bbh'.split()]
    bollband_lo = df_total['time bbl'.split()]
    if not plots:
        global ax, ax2
        fplt.plot(df_total['time'], df_total['Close'].rolling(9).mean(), ax=ax, color='#ff0000', legend='ma-9')
        fplt.plot(df_total['time'], df_total['Close'].rolling(24).mean(), ax=ax, color='#00ff00', legend='ma-24')
        fplt.plot(df_total['time'], df_total['Close'].rolling(100).mean(), ax=ax, color='#0000ff', legend='ma-100')
        fplt.plot(df_total['time'], df_total['Close'].rolling(200).mean(), ax=ax, color='#005555', legend='ma-200')
        plots.append(fplt.candlestick_ochl(candlesticks))
        plots.append(fplt.plot(bollband_hi, color='#4e4ef1'))
        plots.append(fplt.plot(bollband_lo, color='#4e4ef1'))
        #plots.append(fplt.volume_ocv(volumes, ax=ax.overlay()))
        macd = df_total.Close.ewm(span=12).mean() - df_total.Close.ewm(span=26).mean()
        signal = macd.ewm(span=9).mean()
        df_total['macd_diff'] = macd - signal
        fplt.volume_ocv(df_total[['time', 'Open', 'Close', 'macd_diff']], ax=ax2, colorfunc=fplt.strength_colorfilter)
        fplt.plot(macd, ax=ax2, legend='MACD')
        fplt.plot(signal, ax=ax2, legend='Signal')
    else:
        # every time after we just update the data sources on each plot
        plots[0].update_data(df_total)
        plots[1].update_data(bollband_hi)
        plots[2].update_data(bollband_lo)
        #plots[3].update_data(volumes)


fplt.background = '#222222'
fplt.foreground = '#ffffff'
fplt.odd_plot_background = '#222222'
fplt.cross_hair_color = '#00ff00'
fplt.candle_bull_color = '#00ff00'
fplt.candle_bull_body_color = '#007700'
fplt.candle_bear_color = '#ff0000'
fplt.volume_bull_color = fplt.candle_bull_color
fplt.volume_bull_body_color = fplt.candle_bull_body_color
fplt.volume_bear_color = fplt.candle_bear_color
plots = []
# clear the bars.csv
f = open("bars.csv", "w")
f.truncate()
f.close()
ax, ax2 = fplt.create_plot('Realtime ETHUSDT 1m  (Binance)', init_zoom_periods=100, maximize=True, rows=2)
ax.set_visible(crosshair=True, xaxis=True, yaxis=True, xgrid=True, ygrid=True)
get_hist_data = get_bars_historic('ETHUSDT', '1m', '200')
historic_data = get_hist_data[['time', 'Open', 'Close', 'High', 'Low', 'Volume', 'Close_time']]
fplt.timer_callback(update, 5.0)
fplt.show()
```

**Comment by Trembles** [comment, marked as answer]
You can get rid of that wall of numeric conversions in your functions with a one liner,

`data = [pd.to_numeric(i) for i in data]`

And it seems that when you call get_bars() in the update function you aren't doing anything with the returned data? Then you have to open the csv again to get the data you just return with the same function.

And in that save to csv you are saving the same data multiple times?

**Reply by Web3dGuy** [reply]
get_bars_historic() runs once outside of the loop first and it saves the last 200 rows to a csv, from there on in the update function, get_bars() is called and appends the latest 1m candle to the csv. I called it this way just for printing to the console for debugging purposes could have also just done get_bars('ETHUSDT', '1m', '1') stand alone. Anyways the function runs and adds to the csv file properly and every time the script runs the csv is cleared. then the csv is loaded back into a restructured data frame then plotted.  originally I was trying to use pd.append() and pd.concat() to join them but then ran into problems with working with a copy and other stuff, also I read that its not the best way and will get slow fast. csv option was chosen. its a mess, I don't know what I'm doing lol, but its kinda fun. Thanks for the tip, not sure if I even need to do that numeric conversion.

it will be a nice chart to add to my signals dashboard once I figure out all of my gleeps.

**Reply by Trembles** [reply]
But you are doing that conversion with 12 lines of code, either this,
`data = [pd.to_numeric(i) for i in data]`

or all of this,

histdata['Open'] = pd.to_numeric(histdata["Open"])
histdata["Close"] = pd.to_numeric(histdata["Close"])
 histdata['High'] = pd.to_numeric(histdata["High"])
histdata['Low'] = pd.to_numeric(histdata["Low"])
  histdata['Volume'] = pd.to_numeric(histdata["Volume"])
  histdata['qav'] = pd.to_numeric(histdata["qav"])
  histdata['num_trades'] = pd.to_numeric(histdata["num_trades"])
  histdata['taker_base_vol'] = pd.to_numeric(histdata["taker_base_vol"])
  histdata['taker_quote_vol'] = pd.to_numeric(histdata["taker_quote_vol"])
  histdata['ignore'] = pd.to_numeric(histdata["ignore"])

**Reply by Web3dGuy** [reply]
yeah i get that, actually just deleted all of it as it seems unnecessary. lol just disregard that fluff.

```python
def get_bars(symbol, interval, limit):
    url = 'https://api.binance.com/api/v1/klines?symbol=' + symbol + '&interval=' + interval + '&limit=' + limit
    data = json.loads(requests.get(url).text)
    #data = [pd.to_numeric(i) for i in data]
    cur_data = pd.DataFrame(data)
    cur_data.columns = ['time', 'Open', 'High', 'Low', 'Close', 'Volume', 'Close_time',
                        'qav', 'num_trades', 'taker_base_vol', 'taker_quote_vol', 'ignore']
    cur_data['time'] = pd.to_datetime(cur_data['time'], unit='ms', origin='unix')
    cur_data['Close_time'] = pd.to_datetime(cur_data['Close_time'], unit='ms', origin='unix')
    cur_data.to_csv('bars.csv', mode='a', header=False, index=False)
    return cur_data
```

**Reply by highfestiva** [reply]
First of all, reduce code:

```python
def get_bars(symbol, interval, limit, mode='a', header=False):
    url = 'https://api.binance.com/api/v1/klines?symbol=' + symbol + '&interval=' + interval + '&limit=' + limit
    data = json.loads(requests.get(url).text)
    cur_data = pd.DataFrame(data)
    cur_data.columns = ['time', 'Open', 'High', 'Low', 'Close', 'Volume', 'Close_time',
                        'qav', 'num_trades', 'taker_base_vol', 'taker_quote_vol', 'ignore']
    cur_data['time'] = pd.to_datetime(cur_data['time'], unit='ms', origin='unix')
    cur_data['Close_time'] = pd.to_datetime(cur_data['Close_time'], unit='ms', origin='unix')
    cur_data.to_csv('bars.csv', mode=mode, header=header, index=False)
    return cur_data


def get_bars_historic(symbol, interval, limit):
    return get_bars(symbol, interval, limit, mode='w', header=True)
```

Second, call your data a noun and your function a verb. `get_hist_data` is a bad name for a variable, while `hist_data` is better.

Third, the lack of updates you're seeing happens because not all plots are updated, i.e. there is no data to scroll to on the bottom axis when new data appears on the top one.

Third, the zoom bug partly happens because you're lacking data (on the right-hand side of ax2), but also due to data being updated in the top plot, while the bottom blot has no new data (on the left). In the new version you can fix this by:

```python
plots[0].update_data(df_total, gfx=False)
plots[1].update_data(bollband_hi, gfx=False)
...
plots[0].update_gfx()
plots[1].update_gfx()
...

```

**Reply by Web3dGuy** [reply]
Wow, great!  I thought that looked a bit clunky but did not realize you can piggyback through the return like that, TIL.  I cloned the latest finplot and updated the code as per your suggestion but the zooming seems to persist after it updates a few times , and the updated candles still wont display properly. In the updated program the time data the candles and indicators are receiving is the Close_time, as provided from the datastream, I've tried switching it to Open_time but either way the new candles don't display properly, and the indicators don't seem to be tagging along for the ride.  On line 42, what exactly does plt.candlestick_ochl() want/can have for input here?

![Screenshot_2](https://user-images.githubusercontent.com/6704895/109398511-a4faf780-78fa-11eb-9ff6-a365a85da436.png)

I've found that I didn't need to do dateTime conversion and it runs the same, so I removed that, but did notice in the csv that the close time lands on 00:59.999 or rather the epoch time 1614441719999,  is this creating a problem? should it not be rounded up to the next minute?

![Screenshot_3](https://user-images.githubusercontent.com/6704895/109398534-ba702180-78fa-11eb-96e7-69a3d063230c.png)

and lastly I'd like to have the current candle change with the update tick, I understand that because I'm dropping duplicates by Close_time, I only see the candle change every minute, but if I stop dropping duplicates I get "ValueError: cannot reindex from a duplicate axis", meanwhile the csv receives every tick.
updated code:
```python
import pandas as pd
import finplot as fplt
import requests
import json


def get_bars(symbol, interval, limit, mode='a', header=False):
    url = 'https://api.binance.com/api/v1/klines?symbol=' + symbol + '&interval=' + interval + '&limit=' + limit
    data = json.loads(requests.get(url).text)
    cur_data = pd.DataFrame(data)
    cur_data.columns = ['Open_time', 'Open', 'High', 'Low', 'Close', 'Volume', 'Close_time',
                        'qav', 'num_trades', 'taker_base_vol', 'taker_quote_vol', 'ignore']
    cur_data.to_csv('bars.csv', mode=mode, header=header, index=False)
    return cur_data


def get_bars_historic(symbol, interval, limit):
    return get_bars(symbol, interval, limit, mode='w', header=True)


def calc_bollinger_bands(df):
    r = df['Close'].rolling(20)
    df['bbh'] = r.mean() + 2*r.std()
    df['bbl'] = r.mean() - 2*r.std()


def update():
    get_bars('ETHUSDT', '1m', '1')
    df_total = pd.read_csv('bars.csv').drop_duplicates('Close_time').reset_index(drop=True)
    candlesticks = df_total['Close_time Open Close High Low'.split()]
    calc_bollinger_bands(df_total)
    volumes = df_total['Close_time Open Close Volume'.split()]
    bollband_hi = df_total['Close_time bbh'.split()]
    bollband_lo = df_total['Close_time bbl'.split()]

    if not plots:
        global ax, ax2
        fplt.plot(df_total['Close_time'], df_total['Close'].rolling(9).mean(), ax=ax, color='#ff0000', legend='ma-9')
        fplt.plot(df_total['Close_time'], df_total['Close'].rolling(24).mean(), ax=ax, color='#00ff00', legend='ma-24')
        fplt.plot(df_total['Close_time'], df_total['Close'].rolling(100).mean(), ax=ax, color='#0000ff', legend='ma-100')
        fplt.plot(df_total['Close_time'], df_total['Close'].rolling(200).mean(), ax=ax, color='#005555', legend='ma-200')
        plots.append(fplt.candlestick_ochl(candlesticks))
        plots.append(fplt.plot(bollband_hi, color='#4e4ef1'))
        plots.append(fplt.plot(bollband_lo, color='#4e4ef1'))
        plots.append(fplt.volume_ocv(volumes, ax=ax.overlay()))
        macd = df_total.Close.ewm(span=12).mean() - df_total.Close.ewm(span=26).mean()
        signal = macd.ewm(span=9).mean()
        df_total['macd_diff'] = macd - signal
        fplt.volume_ocv(df_total[['Close_time', 'Open', 'Close', 'macd_diff']], ax=ax2, colorfunc=fplt.strength_colorfilter)
        fplt.plot(macd, ax=ax2, legend='MACD')
        fplt.plot(signal, ax=ax2, legend='Signal')
    else:
        # every time after we just update the data sources on each plot
        plots[0].update_data(candlesticks, gfx=False)
        plots[1].update_data(bollband_hi, gfx=False)
        plots[2].update_data(bollband_lo, gfx=False)
        plots[3].update_data(volumes, gfx=False)

        plots[0].update_gfx()
        plots[1].update_gfx()
        plots[2].update_gfx()
        plots[3].update_gfx()

fplt.background = '#222222'
fplt.foreground = '#ffffff'
fplt.odd_plot_background = '#222222'
fplt.cross_hair_color = '#00ff00'
fplt.candle_bull_color = '#00ff00'
fplt.candle_bull_body_color = '#007700'
fplt.candle_bear_color = '#ff0000'
fplt.volume_bull_color = fplt.candle_bull_color
fplt.volume_bull_body_color = fplt.candle_bull_body_color
fplt.volume_bear_color = fplt.candle_bear_color
plots = []
# clear the bars.csv
f = open("bars.csv", "w")
f.truncate()
f.close()
ax, ax2 = fplt.create_plot('Realtime ETHUSDT 1m  (Binance)', init_zoom_periods=100, maximize=True, rows=2)
ax.set_visible(crosshair=True, xaxis=True, yaxis=True, xgrid=True, ygrid=True)
get_bars_historic('ETHUSDT', '1m', '200')
fplt.timer_callback(update, 5.0)
fplt.show()

```
 Thanks for your help man! I'll see if I can get you some coffees or something.

**Reply by highfestiva** [reply]
You should probably use what you call *open time*, which all exchanges simply call *time*.
```python
df_total = pd.read_csv('bars.csv').drop_duplicates('Close_time', keep='last').set_index('Open_time')
```
That way you can drop all the `Close_time`s you were passing into the plot functions. Next you should set
```python
fplt.right_margin_candles = 0
```
in the setup to further avoid some buggy auto-zoom. That seems to make it work. If you like, you could look into using Binance's websocket [API](https://python-binance.readthedocs.io/en/latest/websockets.html), so you won't get the "stop-the-world" updates, but instead a smooth update in 2 Hz or so.

Question: why are you not using tradingview for this?

Good luck with trading!

**Reply by Web3dGuy** [reply]
ok cool, so I added the margin and modified the csv read function with set index but that didn't work, it just gave me a `KeyError: "['Open_time'] not in index"`.  So I tried setting the index in get_bars, also no dice.

what I ended up doing
 In get_bars
` cur_data.to_csv('bars.csv', mode=mode, header=header, index=False)`
 in update
`df_total = pd.read_csv('bars.csv').drop_duplicates('Close_time', keep='last').reset_index(drop=True)` 

and its working! mostly ... it still wants to zoom in periodically and I have to scrollwheel out to see the latest bars, even with the margin set.

why not just TradingView? well I use TradingView but this just seems like a fun way to get familiar with python. I'm going to tinker away until I have a nice little chart and signals dashboard... yes TradingView I know, but this way I can have fun trying some things in python that maybe not so easy just on TradingView.

**Reply by highfestiva** [reply]
I meant plotting like so, without defining x (it's already in the index):
```python
fplt.plot(df_total['Close'].rolling(9).mean(), ax=ax, color='#ff0000', legend='ma-9')
```

**Reply by Web3dGuy** [reply]
if I do that for the MA plots I get 

`line 4307, in __getitem__  result = getitem(key) IndexError: index 4 is out of bounds for axis 0 with size 2`

**Reply by Web3dGuy** [reply]
A bit of fooling around and I realized the I should be using `plots.append(fplt.plot())` instead of` fplt.plot()` alone. The volume overlay seems to get in the way so I put it on a separate axis. I'm not sure why but if I remove the volume plot all together and adjust the plots updates index values, the zoom problem comes back.  one more question , is there an easy way to switch on and off axis visibility? Then i can just turn off volume or MACD via boolean. 

working code:

```python
import pandas as pd
import finplot as fplt
import requests
import json


def get_bars(symbol, interval, limit, mode='a', header=False):
    url = 'https://api.binance.com/api/v1/klines?symbol=' + symbol + '&interval=' + interval + '&limit=' + limit
    data = json.loads(requests.get(url).text)
    cur_data = pd.DataFrame(data)
    cur_data.columns = ['Open_time', 'Open', 'High', 'Low', 'Close', 'Volume', 'Close_time',
                        'qav', 'num_trades', 'taker_base_vol', 'taker_quote_vol', 'ignore']
    cur_data.to_csv('bars.csv', mode=mode, header=header, index=False)
    return cur_data


def get_bars_historic(symbol, interval, limit):
    return get_bars(symbol, interval, limit, mode='w', header=True)


def calc_bollinger_bands(df):
    r = df['Close'].rolling(20)
    df['bbh'] = r.mean() + 2*r.std()
    df['bbl'] = r.mean() - 2*r.std()


def update():
    get_bars('ETHUSDT', '1m', '1')
    df_total = pd.read_csv('bars.csv').drop_duplicates('Close_time', keep='last').reset_index(drop=True)
    candlesticks = df_total['Open_time Open Close High Low'.split()]
    calc_bollinger_bands(df_total)
    volumes = df_total['Open_time Open Close Volume'.split()]
    bollband_hi = df_total['Open_time bbh'.split()]
    bollband_lo = df_total['Open_time bbl'.split()]
    ma9 = df_total['Close'].rolling(9).mean()
    ma24 = df_total['Close'].rolling(24).mean()
    ma100 = df_total['Close'].rolling(100).mean()
    ma200 = df_total['Close'].rolling(200).mean()
    if not plots:
        global ax, ax2, ax3

        plots.append(fplt.candlestick_ochl(candlesticks))
        plots.append(fplt.plot(bollband_hi, color='#4e4ef1'))
        plots.append(fplt.plot(bollband_lo, color='#4e4ef1'))
        plots.append(fplt.volume_ocv(volumes, ax=ax2))
        plots.append(fplt.plot(ma9,ax=ax, color='#ff0000'))
        plots.append(fplt.plot(ma24,ax=ax, color='#00ff00'))
        plots.append(fplt.plot(ma100,ax=ax, color='#0000ff'))
        plots.append(fplt.plot(ma200,ax=ax, color='#005555'))

        macd = df_total.Close.ewm(span=12).mean() - df_total.Close.ewm(span=26).mean()
        signal = macd.ewm(span=9).mean()
        df_total['macd_diff'] = macd - signal
        fplt.volume_ocv(df_total[['Open_time', 'Open', 'Close', 'macd_diff']], ax=ax3, colorfunc=fplt.strength_colorfilter)
        fplt.plot(macd, ax=ax3, legend='MACD')
        fplt.plot(signal, ax=ax3, legend='Signal')
    else:
        # every time after we just update the data sources on each plot
        plots[0].update_data(candlesticks)
        plots[1].update_data(bollband_hi)
        plots[2].update_data(bollband_lo)
        plots[3].update_data(volumes)
        plots[4].update_data(ma9)
        plots[5].update_data(ma24)
        plots[6].update_data(ma100)
        plots[7].update_data(ma200)
        for item in plots:
            item.update_gfx()

fplt.background = '#222222'
fplt.foreground = '#ffffff'
fplt.odd_plot_background = '#222222'
fplt.cross_hair_color = '#00ff00'
fplt.candle_bull_color = '#00ff00'
fplt.candle_bull_body_color = '#007700'
fplt.candle_bear_color = '#ff0000'
fplt.volume_bull_color = fplt.candle_bull_color
fplt.volume_bull_body_color = fplt.candle_bull_body_color
fplt.volume_bear_color = fplt.candle_bear_color
fplt.right_margin_candles = 0

plots = []
# clear the bars.csv
f = open("bars.csv", "w")
f.truncate()
f.close()

ax, ax2, ax3 = fplt.create_plot('Realtime ETHUSDT 1m  (Binance)', init_zoom_periods=100, rows=3)
ax.set_visible(crosshair=True, xaxis=True, yaxis=True, xgrid=True, ygrid=True)
ax2, ax3.set_visible(crosshair=True, xaxis=False, yaxis=True, xgrid=False, ygrid=True)
get_bars_historic('ETHUSDT', '1m', '200')
fplt.timer_callback(update, 5.0)

fplt.show()

```

**Reply by highfestiva** [reply]
Seems that `ax2.hide()` is what you're looking for.

**Reply by highfestiva** [reply]
Thanks for the donation man! Question: I’m thinking of building a bot builder where the user design the algorithm and an optimizer optimizes the parameters and delivers a script for a fee. Do you think there might be a market for such a product if it’s easy to pick an algo that can 10x small amounts in a year?

**Reply by Web3dGuy** [reply]
Yeah, that seems like a pretty good idea, I'd probably put it to use. Normally my strategy is based off of simple stuff like the 100 and 200 moving average along with RSI. For example if the hourly is about to touch off the 200 MA I'll check to see the hourly and 4hr RSI are below at least 40 before assuming a bounce or decent entry. I'll also check for candle compression, and what wave we are on in the trend, if at wave 4 ill start taking some profit and usually be out after wave 5, so some kind of wave counting would be cool. I think it would be cool to have a system where you could for example draw a parallel channel or triangle on a trend and if there is an exit attempt opposite the position by a threshold, consider the RSI and MACD levels on a 2 timeframes before making a decision, and/or send a notification SMS/call for action. Detection of support and resistance areas based on a time area would be cool too, Number of tops bottoms could be counted and considered before doing something stupid. Not sure about letting a bot take control but I am definitely interested in not missing an opportunity, so really smart signals is where I'm at. Regarding the coffee's donation ;) I figured it's well deserved and well.. I'll probably have a few more questions when I get some time to actually look deeper and understand your code, I am also aware that time is not cheap.

**Reply by highfestiva** [reply]
I use a bot which is very trivial, but was able to 13x in a month, which I thought was really good. But it lost it equally quickly. I think in the long run I prefer it doing something stupid from time to time, rather than me making stupid, emotional, irrational decisions all the time. :) I guess most people trading are much better at that, so I'm probably not catering to the bot peoples. What profits are you able to make on what instruments?

**Comment by Web3dGuy** [comment]
I've done pretty good with ETH & ADA this year, up all in all, not by crazy
amounts.  Definitely losing some trades. I don't have the balls to hold for
10x returns. Sometimes the devil gets to me and I play with leverage and
lose lol. Did 300% a couple weeks ago on ETH with 20x, woo big win then two
days later lose all but 20% by missing the chart for a bit. I play for
accumulation these days.  Really regret selling my 5 coins a handful of
years ago, I mean I made good profit but got 100% out of crypto at $3.5k
thinking it needed some downtime before going again.... then it went
straight for 10k , facepalm moment. Always hold, this shits going places!

**Reply by Web3dGuy** [reply]
![received_785238742349359](https://user-images.githubusercontent.com/6704895/109896821-dbb07500-7c4e-11eb-9c9f-a5c2f4ca87ff.jpeg)

This 200ma bounce on btc is a big deal.

**Reply by highfestiva** [reply]
Look into futures, so you're able to short and make money when the price drops. My bot was able to make a 60% gain by shorting BTC@57400 with 7x leverage (but only after a ton of time spent backtesting the algorithm on BTC).
